src/path-graph.js

```diff
@@ -0,0 +1,905 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import { Component } from 'react';
+import { connect } from 'react-redux';
+import * as d3 from 'd3';
+import { faPowerOff } from '@fortawesome/free-solid-svg-icons';
+import { faMapMarkerAlt } from '@fortawesome/free-solid-svg-icons';
+import { faMapMarker } from '@fortawesome/free-solid-svg-icons';
+import { faExpand } from '@fortawesome/free-solid-svg-icons';
+import { faExpandArrowsAlt } from '@fortawesome/free-solid-svg-icons';
+import { faCompressArrowsAlt } from '@fortawesome/free-solid-svg-icons';
+import { faDownload } from '@fortawesome/free-solid-svg-icons';
+
+import { CollapsibleSection } from './collapsible-section.js';
+import { TextButton } from './buttons.js';
+import { downloadSvg } from './util.js';
+import './path-graph.css';
+
+// graph settings
+const minWidth = 240;
+const minHeight = 240;
+const maxWidth = 3840;
+const maxHeight = 2160;
+const minZoom = 0.25;
+const maxZoom = 4;
+const nodeRadius = 25;
+const nodeDistance = 100;
+const nodeCharLimit = 25;
+const nodeFontSize = 8;
+const nodeRepulsion = 200;
+const edgeFontSize = 8;
+const edgeThickness = 2;
+const edgeArrowSize = 10;
+const inkColor = '#424242';
+const backgroundColor = '#fafafa';
+const highlightColor = '#02b3e4';
+
+// path graph section component
+export class PathGraph extends Component {
+  // initialize component
+  constructor() {
+    super();
+
+    this.state = {};
+    this.state.width = maxWidth;
+    this.state.height = maxHeight;
+    this.state.nodeCount = 0;
+    this.state.edgeCount = 0;
+
+    this.graph = React.createRef();
+
+    this.expandContainer = this.expandContainer.bind(this);
+    this.collapseContainer = this.collapseContainer.bind(this);
+    this.updateSectionWidth = this.updateSectionWidth.bind(this);
+    this.setDimensions = this.setDimensions.bind(this);
+    this.setGraphCounts = this.setGraphCounts.bind(this);
+  }
+
+  // when component mounts
+  componentDidMount() {
+    this.updateSectionWidth();
+    this.collapseContainer();
+    window.addEventListener('resize', this.updateSectionWidth);
+  }
+
+  // when component updates
+  componentDidUpdate(prevProps, prevState) {
+    if (this.state.width !== prevState.width) {
+      let width = this.state.width;
+      if (Math.round(width) !== width)
+        width = Math.round(width);
+      if (width > maxWidth)
+        width = maxWidth;
+      if (width < minWidth)
+        width = minWidth;
+      this.setState({ width: width });
```

Comment:
If my memory is correct, the new ES6 or 7 allows you to write this line as:
```this.setState({width});```
It is up to you whether you want to use the simpler syntax. (It did take me a while to get used to it.)

Reply:
I'd prefer to leave it be for now, just because I think there's a bunch of other places in the code where I write it like that and I don't want to go through it all right now.

Perhaps when hetmech is "finished", or in a final draft, we can scan through the whole code to make sure we're using best practices, the cleanest syntax, and things like that.